### ISSUE TITLE
Custom plugin with add_filter fails to activate

### ISSUE THREAD

**Issue by ahmed-sigmalux** [inpsyde/menu-cache#5]
#### Version Information

* PHP: 7.1
* WordPress: 4.8.2

#### Steps to Reproduce

1. Create custom plugin modifying menu caching period
2. Activate custom plugin

```
<?php
add_filter( \Inpsyde\MenuCache\MenuCache::FILTER_EXPIRATION, function () {
  return 3600;
} );
?>
```

#### What I Expected

Custom plugin would activate and filter would modify menu caching period.

#### What Happened Instead

Custom plugin fails to activate:

```
[27-Sep-2017 05:14:53 UTC] PHP Fatal error:  Uncaught Error: Class 'Inpsyde\MenuCache\MenuCache' not found in /var/www/mysite.com/wp-content/plugins/customplugin-menu-cache/customplugin-menu-cache.php:14
Stack trace:
#0 /var/www/mysite.com/wp-admin/includes/plugin.php(560): include_once()
#1 /var/www/mysite.com/wp-admin/plugins.php(42): activate_plugin('customplugin-menu-...', 'https://www.mysite.com...', true)
#2 /var/www/mysite.com/wp-admin/network/plugins.php(13): require('/var/www/www.mysite.com...')
#3 {main}
  thrown in /var/www/mysite.com/wp-content/plugins/customplugin-menu-cache/customplugin-menu-cache.php on line 14
```

**Comment by szepeviktor** [comment]
Using a constant from another plugin as a filter name is rather unfortunate as it may not be loaded yet.

Try delaying add_filter until init.

-- 
+36204242498
Ezen a készüléken sok az elütés. 
Elnézést!

**Comment by ahmed-sigmalux** [comment]
I've tried changing the priority to 99 and 999 without success...

**Comment by szepeviktor** [comment]
The nicest place to put this code is an MU plugin.

`/var/www/mysite.com/wp-content/mu-plugins/menu-cache-filter.php`

```php
<?php
add_action( 'init', function () {
  add_filter( \Inpsyde\MenuCache\MenuCache::FILTER_EXPIRATION, function () {
    return 3600;
  } );
} );
```

Or you may go `'plugins_loaded'` with priority higher than 10, see https://github.com/inpsyde/menu-cache/blob/master/index.php#L46

BTW I think Inpsyde uses this plugin in custom themes.

**Comment by gmazzap** [comment]
No, we don't use plugins from themes.

But @szepeviktor is right, @ahmed-sigmalux  to use anything from a plugin you need to be sure the plugin is available.

Also, to use a class constant you need to be sure the class is loaded. And in a place where autoload is in place, the class is loaded the first time is instantiated or referenced in code.

For the class `\Inpsyde\MenuCache\MenuCache` that happens at `"plugins_loaded"` with priority 10.

Which means that you could use same hook but with higher priority, e.g. 

```php
add_action( 'plugins_loaded', function () {
  if ( ! class_exists( 'Inpsyde\MenuCache\MenuCache' ) ) {
      return;
  }
  add_filter( \Inpsyde\MenuCache\MenuCache::FILTER_EXPIRATION, function () {
    return HOUR_IN_SECONDS;
  } );
}, 20 );
```

The snippet above can be used either from a plugin or from a MU plugin.

If you want to add it from a theme, there's no need for the "wrapping" hook, because when theme is loaded all plugins are loaded already.

A snippet like: 

```php
if ( class_exists( 'Inpsyde\MenuCache\MenuCache' ) ) {
    add_filter( \Inpsyde\MenuCache\MenuCache::FILTER_EXPIRATION, function () {
        return HOUR_IN_SECONDS;
    } );
}
```

can go to `functions.php` of a theme.

**Comment by ahmed-sigmalux** [comment]
Thank you, @szepeviktor and @gmazzap. Your posts explained things clearly and I now have a properly functioning custom plugin.

**Comment by gmazzap** [comment]
Perfect @ahmed-sigmalux I will close the issue. And thanks to @szepeviktor for helping here.

**Comment by bueltge** [comment]
@gmazzap Maybe we should add this to the wiki for users, there have the same topic?

**Comment by tfrommen** [comment]
FYI: I just released a new version where the plugin gets bootstrapped earlier, with priority `0`.

As long as your code runs after this (e.g., just hooked to `plugins_loade`, without a priority specified) this should do just fine.

Regarding how and when to do your customization, all important things have been mentioned already.

Preferably, use a custom (MU) plugin, or put the customization in your theme's `functions.php` file. If in a plugin, make it run after the plugin has been loaded; well, and before the menu gets rendered. 😉 And if you want to use the class constants, make sure the class is available. I added a short note in the readme.

Cheers,
Thorsten
